include/arch/arm/arch/object/vcpu.h

```diff
@@ -76,10 +79,10 @@ struct vcpu {
     bool_t vppi_masked[n_VPPIEventIRQ];
 #ifdef CONFIG_VTIMER_UPDATE_VOFFSET
     /* vTimer is 8-bytes wide and has same alignment requirement.
-     * To keep the struct packed on 32-bit platforms when accompanied by an
-     * odd number of 32-bit words, we need to add a padding word.
+     * The struct will remain packed on 32-bit platforms when n_VPPIEventIRQ
```

Comment:
Well, it's currently packed. If `n_VPPIEventIRQ` becomes even for whatever reason, it'll stop being packed on aarch32, and you'll need to add an extra word.
I'm not sure we're talking about the same thing. A packed struct has no implicit padding. "Generate the padding automatically" means the compiler will add implicit padding, which is no good for verification.

Reply:
Maybe since there's no padding there now, I should remove the comment if it's causing confusion.

Reply:
I think the comment should just be written in a more clear way, e.g. that we want no padding for verification, but don't want toi declare the struct as packed explicitly.

Seems we cold also use this safe guard to get a compile error if padding is to be added:
```
#pragma GCC diagnostic push
#pragma GCC diagnostic error "-Wpadding"
.... struct(s) that must not have padding...
#pragma GCC diagnostic pop
```

And for clang is seems there is "#pragma clang diagnostic ..."

As a side note, isn't "bool_t vppi_masked[n_VPPIEventIRQ];" slightly broken anyway, as this is not a bit-field, but use one full `int` per bit. But as long as there are just a few elements that seems acceptable for now.

Reply:
> I think the comment should just be written in a more clear way, e.g. that we want no padding for verification, but don't want toi declare the struct as packed explicitly.

What do you mean "declare the struct as packed explicitly"? Do you know what the `__packed__` attribute does? It's not what we want. We want packed, aligned, and understandable by the C parser.

Also, that is not more clearly describing what's going on, it just confuses matters further. As I wrote, if a struct has no padding, it's packed, whether you "declare" it as such or not. The comment is a warning that if you make the array one size bigger, you'll need padding on 32-bit.

> Seems we cold also use this safe guard to get a compile error if padding is to be added:
> 
> ```
> #pragma GCC diagnostic push
> #pragma GCC diagnostic error "-Wpadding"
> .... struct(s) that must not have padding...
> #pragma GCC diagnostic pop
> ```
> 
> And for clang is seems there is "#pragma clang diagnostic ..."

Two different pragmas for two different compilers. What's the guideline for this PR? *All* structures we deal with in verification must be packed, only the ones outside the verified subset. Also your suggestions would only generate an error when combined with "-Werror". This is going in an unproductive direction.

> As a side note, isn't "bool_t vppi_masked[n_VPPIEventIRQ];" slightly broken anyway, as this is not a bit-field, but use one full `int` per bit. But as long as there are just a few elements that seems acceptable for now.

That's not broken, that's locally inefficient. It doesn't matter because VCPUs occupy a page in memory anyway.

Reply:
I think the fundamental misunderstand here is that verification, as Raf points out, does require *all* structs to contain no padding (this is to avoid unspecified memory content in the padding fields). At least all structs where verification needs to reason about field access -- there are some (very) few where that isn't necessary.

This requirement does not seem to be documented anywhere currently. It doesn't make sense to add a comment on this to one particular struct, but we could add it to the [verification requirements in the style guide](https://docs.sel4.systems/processes/style-guide.html).

With this knowledge as background I see no problem with the comment.

Reply:
See now https://github.com/seL4/docs/pull/202

Reply:
I'd support adding it to the style/coding conventions. It would also be nice to tell before we even import the code into Isabelle what's packed and what isn't, but we only really see this come up when adding new structures, expanding the verification scope and/or verifying new arches so I guess we can keep things as they are for a time.